keps/sig-api-machinery/storage-migration.md

```diff
@@ -81,158 +81,109 @@ As to the deliverables,
+We will support a `kube-migrator migrate --resources=<list of resources>`
+command. It (re)creates `migration` CRs for the resources to trigger their
+migration. This command facilitates manually migrating custom resources and
+aggregated resources.
 
-### API
+### Alpha API
 
-We introduce the `migration` API to record the intention and progress of a
+We introduce the `migration` API to record the intention and the progress of a
 migration. The API will be a CRD defined in the `migration.k8s.io` group.
 
 Read the [workflow section][] to understand how the API is used.
 
 ```golang
 type StorageVersionMigration struct {
   metav1.TypeMeta
+  // metadata.name should be "<resource>.<group>" of the resource being
+  // migrated.
+  // Because resouces can be served at different groups, in the future version,
+  // metadata.name can be hash(etcd instance, etcd path prefix).
   metav1.ObjectMeta
   Spec StorageVersionMigrationSpec
   Status StorageVersionMigrationStatus
 }
 
-// Note that the apiserver will increment `metadata.generation` when this spec
-// is updated.
 type StorageVersionMigrationSpec {
-  // Resource is the identifier of the resource that's being migrated.
-  // TBD: maybe resource=hash(etcd instance, etcd path prefix).
-  // Immuntable.
-  Resource string 
-
-  // GroupVersions are all the groups and versions the apiserver serves this
-  // resource in. The server preferred version should be groupVersions[0].
-  // **Migration worker** GET & PUT via the preferred version by default, to
-  // avoid conversions in the apiserver.
-  GroupVersions []GroupVersion
-  
-  // LastUpdateTimestamp is the last time this spec is updated. The value is set
-  // by clients, not managed by the apiserver. The timestamp introduces randomness
-  // to the spec so that every update of the spec changes the contents of the spec,
-  // which leads to the increment of `metadata.generation`.
-  LastUpdateTimestamp metav1.Time
-
-  // Reason of the migration. Examples are "manually triggered", "apiserver
-  // updated", etc.
-  Reason string
-}
-
-type GroupVersion struct {
-  Group string
-  Version string
+  // Resource is the resource that is being migrated. The migrator sends
+  // requests to the endpoint serving the Resource.
+  // Immutable.
+  Resource GroupVersionResource
 }
 
 type StorageVersionMigrationStatus {
-  // ObeservedGeneration is the spec generation that has been observed by the
-  // **migration controller**. It indicates a **migration worker** for this
-  // generation has been launched.
-  ObeservedGeneration int64
-
-  // CompletedGeneration is the greatest spec generation whose **migration worker**
-  // has completed.
-  CompletedGeneration int64
+  // Completed means the migration of this resource has completed.
+  Completed bool
+
+  // ContinueToken is the token to use in the list options to get the next chunk
+  // of objects to migrate. User can use this token to check the progress of the
```

Comment:
Can the user use this to tell if the migration has started or not? If yes, how? If no, can we add this to the API?

Reply:
It would be empty if the migration hasn't started. And if I added `.status.active` to `migration` as in https://github.com/kubernetes/community/pull/2524#discussion_r221792276, that would indicate the start as well.

Reply:
Don't add a bunch of bools if an enum will serve.

Also for this one respond in the comment (responding on github is optional) ;)